native/src/components/PinchableBox.tsx

```diff
@@ -0,0 +1,317 @@
+    const heightIncreaseByScale = (Math.max(0, this.lastScale - 1) * realImageHeight) / 2
+    const minX = 0 - widthIncreaseByScale
+    const maxX = viewWidth + widthIncreaseByScale - viewWidth
+    const minY = 0 - heightIncreaseByScale
+    const maxY = viewHeight + heightIncreaseByScale - viewHeight
+
+    if (newX <= minX) {
+      this.fixBound('x', minX, newX)
+    }
+
+    if (newX >= maxX) {
+      this.fixBound('x', maxY, newX)
+    }
+
+    if (newY <= minY) {
+      this.fixBound('y', minY, newY)
+    }
+
+    if (newY >= maxY) {
+      this.fixBound('y', maxY, newY)
+    }
+
+    if (newScale <= 1) {
+      // Disable gesture handler during animation
+      this.setState(state => ({ ...state, interactive: false }))
+
+      const animation: Animated.CompositeAnimation = Animated.timing(this.baseScale, {
+        toValue: 1,
+        duration: ANIMATION_DURATION,
+        useNativeDriver: USE_NATIVE_DRIVER
+      })
+      animation.start(({ finished }) => {
+        if (finished) {
+          this.lastScale = 1
+        }
+        this.setState(state => ({ ...state, interactive: true }))
+      })
+    }
+  }
+
+  private createOnPinchHandlerStateChange =
+    (viewWidth: number, viewHeight: number, realImageWidth: number, realImageHeight: number) =>
+    (event: PinchGestureHandlerStateChangeEvent) => {
+      // If the two fingers were lifted from the screen, then...
+      if (event.nativeEvent.oldState === State.ACTIVE) {
+        const scaleDelta = this.lastScale
+        const newScale = scaleDelta * event.nativeEvent.scale
+        this.lastScale = newScale
+        this.baseScale.setValue(newScale)
+        this.pinchScale.setValue(1)
+
+        this.fixBounds(
+          viewWidth,
+          viewHeight,
+          realImageWidth,
+          realImageHeight,
+          this.lastOffset.x,
+          this.lastOffset.y,
+          this.lastScale
+        )
+      }
+    }
+
+  private createOnPanHandlerStateChange =
+    (viewWidth: number, viewHeight: number, realImageWidth: number, realImageHeight: number) =>
+    (event: PanGestureHandlerStateChangeEvent) => {
+      if (event.nativeEvent.state === State.END) {
+        // Distances the finger was dragged across the screen
+        const movedInY = event.nativeEvent.translationY
+        const movedInX = event.nativeEvent.translationX
+        const previousY = this.lastOffset.y
+        const previousX = this.lastOffset.x
+
+        const newY = previousY + movedInY
+        this.lastOffset.y = newY
+        this.translateY.setOffset(newY)
+        this.translateY.setValue(0)
+
+        const newX = previousX + movedInX
+        this.lastOffset.x = newX
+        this.translateX.setOffset(newX)
+        this.translateX.setValue(0)
+
+        this.fixBounds(viewWidth, viewHeight, realImageWidth, realImageHeight, newX, newY, this.lastScale)
+      }
+    }
+
+  render(): ReactElement | null {
+    const pinchHandler = React.createRef()
```

Comment:
i think best practise is creating refs once in the constructor if you use class components